neonvm/runner/main.go

```diff
@@ -532,10 +544,198 @@ func main() {
 	}
 
-	go terminateQemuOnSigterm(vmSpec.QMP)
-	if err := execFg(QEMU_BIN, qemuCmd...); err != nil {
+	isSystemd, err := isSystemdCgroup()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var cgroupPath string
+	cgroupName := fmt.Sprintf("%s-vm-runner", vmStatus.PodName)
+	if isSystemd {
+		cgroupPath = path.Join(cgroupSystemdBase, cgroupName)
+	} else {
+		cgroupPath = path.Join(cgroupPlainBase, cgroupName)
+	}
+
+	if err := createCgroup(cgroupPath); err != nil {
+		log.Fatal(err)
+	}
+
+	if err := setCgroupLimit(qemuCPUs.minFraction, cgroupPath); err != nil {
 		log.Fatal(err)
 	}
+
+	go terminateQemuOnSigterm(vmSpec.QMP)
+	go listenForCPUChanges(vmSpec.RunnerPort, cgroupPath)
+
+	qemuString := shellescape.QuoteCommand(append([]string{QEMU_BIN}, qemuCmd...))
+	cgropedQemu := fmt.Sprintf("echo $$ >> %s && %s", path.Join(cgroupPath, "cgroup.procs"), qemuString)
+	// start a new shell, cgroup it and then start qemu. Qemu will get cgroup from the parent sh while runner is not cgrouped
+	if err := execFg("sh", "-c", cgropedQemu); err != nil {
+		log.Println(err)
+	}
+
+	log.Println("cleaning up cgroup...")
+	err = os.Remove(cgroupPath)
+	if err != nil {
+		log.Println(err)
+	}
+	log.Println("exiting")
+}
+
+func handleCPUChange(w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		log.Printf("unexpected method: %s\n", r.Method)
+		w.WriteHeader(500)
+		return
+	}
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Printf("could not read body: %s\n", err)
+		w.WriteHeader(500)
+		return
+	}
+
+	parsed := neonApi.VCPUChange{}
+	err = json.Unmarshal(body, &parsed)
+	if err != nil {
+		log.Printf("could not parse body: %s\n", err)
+		w.WriteHeader(500)
+		return
+	}
+
+	cgroupPath := r.Context().Value(cgroupContextKey{})
+	val, ok := cgroupPath.(string)
+	if !ok {
+		log.Println("no path in context")
+		w.WriteHeader(500)
+		return
+	}
+
+	// update cgroup
+	fraction := getCPUFraction(parsed.VCPUs)
+	log.Printf("got CPU update %v (%v)\n", parsed.VCPUs, fraction)
+	err = setCgroupLimit(fraction, val)
+	if err != nil {
+		log.Printf("could not read body: %s\n", err)
+		w.WriteHeader(500)
+		return
+	}
+
+	w.WriteHeader(200)
+}
+
+type cgroupContextKey struct{}
+
+func listenForCPUChanges(port int32, cgroupPath string) {
+	mux := http.NewServeMux()
+	mux.HandleFunc("/cpu_change", handleCPUChange)
+	baseCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	baseCtx = context.WithValue(baseCtx, cgroupContextKey{}, cgroupPath)
+	server := http.Server{
+		Addr:              fmt.Sprintf("0.0.0.0:%d", port),
+		Handler:           mux,
+		ReadTimeout:       5 * time.Second,
+		ReadHeaderTimeout: 5 * time.Second,
+		WriteTimeout:      5 * time.Second,
+		BaseContext: func(_ net.Listener) context.Context {
+			return baseCtx
+		},
+	}
+	err := server.ListenAndServe()
+	if errors.Is(err, http.ErrServerClosed) {
+		log.Println("server closed")
+	} else if err != nil {
+		log.Fatalf("error starting server: %s\n", err)
```

Comment:
I'm pretty sure that the log helpers add a `\n` at the end.

Reply:
I tried to not use them in new code but there may be some places left

Reply:
linking my review from a few days ago, re: this topic: https://github.com/neondatabase/autoscaling/pull/172/files#r1175697800

> This has the trailing `\n` but honestly I think it's mostly fine. There's existing code that has it, so a full cleanup can be done in a separate PR.